pkg/opsgenieIntegrationManager/client.go

```
@@ -0,0 +1,106 @@
+	"github.com/atlassian/voyager/pkg/util/pkiutil"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	asapAudience     = "micros-server"
+	asapSubject      = ""
+	integrationsPath = "/api/v1/opsgenie/integrations"
+)
+
+type Client struct {
+	logger     *zap.Logger
+	httpClient *http.Client
+	asap       pkiutil.ASAP
+	rm         *restclient.RequestMutator
+}
+
+func New(logger *zap.Logger, httpClient *http.Client, asap pkiutil.ASAP, baseURL *url.URL) *Client {
+	rm := restclient.NewRequestMutator(
+		restclient.BaseURL(baseURL.String()),
+	)
+	return &Client{
+		logger:     logger,
+		httpClient: httpClient,
+		asap:       asap,
+		rm:         rm,
+	}
+}
+
+// Gets OpsGenie integrations
+// return codes:
+// - 201: Service created
+// - 400: Bad request to Opsgenie
+// - 401: Unauthorized
+// - 404: Not found returned by Opsgenie. Does the specified Opsgenie team exist?
+// - 422: Semantic error in request to Opsgenie.
+// - 429: Rate limited by Opsgenie.
+// - 500: Internal server error
+func (c *Client) GetOrCreateIntegrations(ctx context.Context, teamName string) (bool /* retriable */, *IntegrationsResponse, error) {
+	req, err := c.rm.NewRequest(
+		pkiutil.AuthenticateWithASAP(c.asap, asapAudience, asapSubject),
+		restclient.Method(http.MethodGet),
+		restclient.JoinPath(fmt.Sprintf("%s/%s", integrationsPath, teamName)),
+		restclient.Context(ctx),
+	)
+	if err != nil {
+		return false, nil, errors.Wrap(err, "failed to create get integrations request")
+	}
+
+	response, err := c.httpClient.Do(req)
+	if err != nil {
+		return false, nil, errors.Wrap(err, "failed to execute get integrations request")
+	}
+
+	retriable := false
+	switch response.StatusCode {
+	case http.StatusInternalServerError:
+		retriable = true
+	case 429:
+		retriable = true
+	}
+
+	defer util.CloseSilently(response.Body)
+	respBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return retriable, nil, errors.Wrap(err, "failed to read response body")
```

also not sure if `ioutil.ReadAll()` returns error on empty body? if it does, we should probably swallow the error for non-200 response

---

I used the error tests found in the PR to validate the error returned is not an ioutil error. i.e. ioutil.ReadAll does not return an error on empty body.